core/parser.mly

```diff
@@ -929,22 +929,32 @@ forall_datatype:
      S = !(ModuleA.ModuleB.Type).S
 
      Parenthesised versions take priority over non-parenthesised versions.
+
+   jcheney: Modified grammar to allow unparenthesized qualified type names
+   in most places, but specifically forbid in ? . or ! .
+   This also requires moving the TILDE production into primary_datatype
+   since otherwise, ? ~M.x . y is ambiguous.
+   This is not ideal since it spreads the session-related constructs
+   among several nonterminals.
 */
 session_datatype:
-| BANG datatype DOT datatype                                   { `Output ($2, $4) }
-| QUESTION datatype DOT datatype                               { `Input  ($2, $4) }
+| BANG primary_datatype_loc DOT datatype                       { `Output ($2, $4) }
+| QUESTION primary_datatype_loc DOT datatype                   { `Input  ($2, $4) }
 | LBRACKETPLUSBAR row BARPLUSRBRACKET                          { `Select $2       }
 | LBRACKETAMPBAR row BARAMPRBRACKET                            { `Choice $2       }
-| TILDE datatype                                               { `Dual $2         }
 | END                                                          { `End             }
 | primary_datatype                                             { $1               }
+| qualified_type_name                                          { `QualifiedTypeApplication ($1, []) }
 
 parenthesized_datatypes:
 | LPAREN RPAREN                                                { [] }
-| LPAREN qualified_type_name RPAREN                            { [with_pos $loc (`QualifiedTypeApplication ($2, []))] }
 | LPAREN datatypes RPAREN                                      { $2 }
 
+primary_datatype_loc:
```

Comment:
Nitpick: across the compiler we use the term "position" to refer to a token's placement in a source file, but "location" seems to be a more commonly used name in the literature (e.g. Menhir uses "location" term). That being said, it would be nice to be consistent about naming and thus this production would probably be better named as `primary_datatype_pos`.

Reply:
OK with me, I wasn't sure if there was a convention already for this.